src/v/iceberg/merge_append_action.cc

```diff
@@ -0,0 +1,543 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/merge_append_action.h"
+
+#include "base/units.h"
+#include "base/vlog.h"
+#include "iceberg/logger.h"
+#include "iceberg/manifest.h"
+#include "iceberg/manifest_file_packer.h"
+#include "iceberg/manifest_list.h"
+#include "iceberg/snapshot.h"
+#include "iceberg/table_requirement.h"
+#include "iceberg/values_bytes.h"
+#include "random/generators.h"
+
+#include <iterator>
+#include <limits>
+
+namespace iceberg {
+
+namespace {
+manifest_path get_manifest_path(
+  const ss::sstring& location, const uuid_t& commit_uuid, size_t num) {
+    return manifest_path{
+      fmt::format("{}/metadata/{}-m{}.avro", location, commit_uuid, num)};
+}
+manifest_list_path get_manifest_list_path(
+  const ss::sstring& location,
+  snapshot_id snap_id,
+  const uuid_t& commit_uuid,
+  size_t num) {
+    return manifest_path{fmt::format(
+      "{}/metadata/snap-{}-{}-{}.avro", location, snap_id(), commit_uuid, num)};
+}
+
+action::errc to_action_errc(manifest_io::errc e) {
+    switch (e) {
+    case manifest_io::errc::failed:
+        return action::errc::io_failed;
+    case manifest_io::errc::shutting_down:
+        return action::errc::shutting_down;
+    case manifest_io::errc::timedout:
+        // NOTE: treat IO timeouts the same as other IO failures.
+        // TODO: build out retry logic.
+        return action::errc::io_failed;
+    }
+}
+
+snapshot_id random_snap_id() {
+    return snapshot_id{random_generators::get_int<int64_t>(
+      0, std::numeric_limits<int64_t>::max())};
+}
+
+snapshot_id generate_unused_snap_id(const table_metadata& m) {
+    auto sid = random_snap_id();
+    if (!m.snapshots.has_value() || m.snapshots->empty()) {
+        return sid;
+    }
+    // Repeatedly try to generate a new snapshot id that isn't used already.
+    const auto& snaps = *m.snapshots;
+    while (std::ranges::find(snaps, sid, &snapshot::id) != snaps.end()) {
```

Comment:
just thinking here: it's weird that the projection needs to know the type (e.g. `snapshot`). if `snapshot` has `operator==(snapshot_id)` then would plain `std::find(snaps, sid)` work?

Reply:
I just tried it out with schemas and I don't think this works (at least the simple/obvious implementations didn't work)